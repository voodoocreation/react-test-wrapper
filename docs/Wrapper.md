`Wrapper`
=========

This is the base class which is used for basic presentational components. The other classes provided
extend this one, so all methods defined on this class are also available on the others.

As with all of these classes, you can extend it to implement your own additional functionality.

Public read-only properties
---------------------------

### `props`
The props the component was mounted with. Useful for avoiding needing to declare local variables
for later assertion reference - especially mock function instances.


Public methods
--------------

### `withDefaultChildren`
Sets the default children to be used for the wrapper instance.

### `withDefaultProps`
Sets the default props to be used for the wrapper instance.

### `withChildren`
Sets the scenario-specific children to be used - cleared after `mount`, `render` or `shallow` are called.

### `withProps`
Sets the scenario-specific props to be used - cleared after `mount`, `render` or `shallow` are called.

### `mount`
Mounts the component with the Enzyme `mount` function, using the currently-set data.
Returns a `ReactWrapper` instance.

### `render`
Mounts the component with the Enzyme `render` function, using the currently-set data.
Returns a `ShallowWrapper` instance.

### `shallow`
Mounts the component with the Enzyme `shallow` function, using the currently-set data.
Returns a `Cheerio` instance.

How to extend
-------------

To extend this class to implement your own setup functionality, you can do so as shown in the
example below.

There's a `WrappingComponent` property on the `Wrapper` class that will automatically be used by
`mount`, `render` and `shallow` when it's defined, to wrap the component being tested.
This is very useful when testing components that require some form of context provider component to
exist in the React tree.

```typescript jsx
import * as React from "react";
import { Wrapper as BaseWrapper } from "react-test-wrapper";

export class WrapperWithCustomStuff<
  C extends React.ComponentType<any>,
  P extends React.ComponentProps<C> = React.ComponentProps<C>
> extends BaseWrapper<C, P> {
  protected WrappingComponent: React.FC = ({ children }) => (
    <SomeProviderComponent>
      {children}
    </SomeProviderComponent>
  );

  // Add custom properties and methods here
}
```
