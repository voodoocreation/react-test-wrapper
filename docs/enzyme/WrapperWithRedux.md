`WrapperWithRedux`
=================

This abstract class has a `WrappingComponent` defined which wraps the component in a Redux `Provider`,
passing in the return value from the `createStore` method.

To use this class, you have to extend it, pass in your store state type in the class declaration and
define your own `createStore` method to return an instance of your Redux store, using the
`initialState` and `middlewares` that this class provides - it is important to ensure that your
store uses these, because if it disregards them, none of the methods this class provides will function.

Extensions to the returned `ReduxWrapper`
-----------------------------------------

The return type of the `mount` method extends Enzyme's `ReactWrapper` by adding a store property to
it, so you can access the store instance that the component was mounted with. This is useful for
some edge cases where you may want to test how your component reacts to actions being dispatched
outside of the component's scope.

For example:
```typescript jsx
const component = new WrapperWithRedux(SomeComponent);

describe("when testing a scenario", () => {
  const wrapper = component
    .withReduxState({
      test: {
        value: "Scenario value 1"
      }
    })
    .mount();

  it("renders the initial value", () => {
    expect(wrapper.find(".SomeComponent--value").text()).toBe("Initial value");
  });

  it("dispatches actions.setValue", () => {
    wrapper.store.dispatch(actions.setValue("New value"));
  });

  it("renders the updated value", () => {
    expect(wrapper.find(".SomeComponent--value").text()).toBe("New value");
  });
});
```


Public read-only properties
---------------------------

### `reduxHistory`
An array of actions that have been dispatched, used when asserting that actions have been
dispatched as expected during interactions or in the component lifecycle.


Public methods
--------------

In addition to the public methods provided by the base [`Wrapper`](Wrapper.md) class, the following
methods are available.

Note:
While technically the `render` and `shallow` methods are accessible, they will throw an error
due to issues with the integration of the Redux store provider.

`render` doesn't work due to its use of `useLayout`, which doesn't work in a test environment.
`shallow` isn't useful due to it only mounting the Redux provider component.

### `withDefaultReduxState`
Sets the default Redux store state to be used for the wrapper instance.

### `withReduxState`
Sets the scenario-specific Redux store state to be used - cleared after `mount`, `render` or `shallow` are called.

### `withMergedReduxArrays`
Toggles whether arrays get merged or not in Redux state.

### `mount`
Mounts the component with the Enzyme `mount` function, using the currently-set data.
Returns a `ReactWrapper` instance, which also includes a `store` property.


How to extend for use in your project
-------------------------------------

```typescript jsx
import * as React from "react";
import { WrapperWithRedux as BaseWrapper } from "react-test-wrapper/enzyme";

import { createStore, TStoreState } from "../store";

class Wrapper<
  C extends React.ComponentType<any>,
  S extends {} = TStoreState,
  P extends React.ComponentProps<C> = React.ComponentProps<C>
> extends BaseWrapper<C, S, P> {
  protected createStore(
    initialState: DeepPartial<S>,
    middlewares: Middleware[]
  ) {
    return createStore(initialState, middlewares);
  }
}
```
