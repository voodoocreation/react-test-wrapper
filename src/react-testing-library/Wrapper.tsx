import {
  cleanup,
  render,
  queries as defaultQueries,
  RenderResult,
} from "@testing-library/react";
import * as React from "react";

import { queries as customQueries } from "./queries.js";

const queries = {
  ...defaultQueries,
  ...customQueries,
};

export type TRenderResult<P> = RenderResult<typeof queries> & {
  updateProps: (newProps: Partial<P>) => void;
};

/**
 * A class to provide a simple interface for setting up your React component unit tests
 */
export class Wrapper<
  C extends React.ComponentType<any>,
  P extends React.ComponentProps<C> = React.ComponentProps<C>,
> {
  // region Properties that are implemented in classes that extend Wrapper

  /**
   * The component that will wrap the component you're testing.
   *
   * This is useful adding provider components to wrap your components with.
   */
  protected WrappingComponent: React.ComponentType<any> | undefined = undefined;

  // endregion

  // region Properties that remain throughout instance lifecycle

  /**
   * The default children for all test scenarios for the current test suite.
   *
   * This is set by the `.withDefaultChildren()` method.
   */
  protected defaultChildren: React.ReactNode;

  /**
   * The default props for all test scenarios for the current test suite.
   *
   * This is set by the `.withDefaultProps()` method.
   */
  protected defaultProps: Partial<P> = {};

  // endregion

  // region Scenario-specific properties that are cleared whenever shallow/mount/render are called

  /**
   * The children specific to the test scenario.
   *
   * This is set by the `.withChildren()` method and cleared after `.mount()`, `.render()` or `.shallow()` are called.
   */
  protected scenarioChildren: React.ReactNode;

  /**
   * The props specific to the test scenario.
   *
   * This is set by the `.withProps()` method and cleared after `.mount()`, `.render()` or `.shallow()` are called.
   */
  protected scenarioProps: Partial<P> = {};

  // endregion

  // region Properties to be accessed via getters after mounting

  /**
   * The merged props to be used during the lifecycle of the test scenario.
   *
   * This is set by the `.mount()`, `.render()` and `.shallow()` methods so that you can access the
   * props in your test via the `.props` getter.
   */
  protected mergedProps: Partial<P> = {};

  // endregion

  /**
   * @param Component The React component to test
   */
  constructor(protected readonly Component: C) {}

  /**
   * Returns the props for the current test scenario lifecycle
   */
  public get props() {
    return this.mergedProps;
  }

  /**
   * Sets the default children to be used in all test scenarios for the current test suite.
   *
   * This method can be chained with other methods.
   *
   * @param children  The children to set
   */
  public withDefaultChildren = (children: React.ReactNode) => {
    this.defaultChildren = children;

    return this;
  };

  /**
   * Sets the default props to be used in all test scenarios for the current test suite.
   *
   * This method can be chained with other methods.
   *
   * @param props The props to set
   */
  public withDefaultProps = (props: Partial<P>) => {
    this.defaultProps = props;

    return this;
  };

  /**
   * Sets the children to be used by the current test scenario.
   *
   * This method can be chained with other methods.
   *
   * @param children  The children to set
   */
  public withChildren = (children: React.ReactNode) => {
    this.scenarioChildren = children;

    return this;
  };

  /**
   * Sets the props to be used by the current test scenario.
   *
   * This method can be chained with other methods.
   *
   * @param props The props to set
   */
  public withProps = (props: Partial<P>) => {
    this.scenarioProps = props;

    return this;
  };

  /**
   * Renders the component using the `render` function from `react-testing-library`.
   *
   * Returns the `RenderResult` from `render()`.
   */
  public render = (): TRenderResult<P> => {
    this.beforeRender();

    const props = this.defineProps() as P;
    const result = render<typeof queries>(<this.Component {...props} />, {
      queries,
      wrapper: this.WrappingComponent,
    });

    this.reset();

    const updateProps = (newProps: Partial<P>) => {
      const merged: P = { ...props, ...newProps };

      result.rerender(<this.Component {...merged} />);
    };

    return {
      ...result,
      updateProps,
    };
  };

  /**
   * This is called before `.render()` is called.
   *
   * You can use this to perform additional functionality when extending this class.
   */
  protected beforeRender = () => {
    // Implement this method when extending to define properties to pass to `WrappingComponent`
    cleanup();
  };

  /**
   * Defines the merged props for the current test scenario lifecycle
   */
  protected defineProps = () => {
    this.mergedProps = {
      ...this.defaultProps,
      ...this.scenarioProps,
      children: this.scenarioChildren || this.defaultChildren,
    };

    return this.mergedProps;
  };

  /**
   * Resets the scenario-specific children and props for the instance
   */
  // eslint-disable-next-line
  protected reset() {
    this.scenarioChildren = undefined;
    this.scenarioProps = {};
  }
}
