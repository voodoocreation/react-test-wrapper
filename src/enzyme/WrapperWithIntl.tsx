import * as React from "react";
import { IntlConfig, IntlProvider } from "react-intl";

import Wrapper from "./Wrapper";

/**
 * A class to provide a simple interface for setting up your React component unit tests, with support for `react-intl`.
 *
 * This class shouldn't be used when using `react-intl-redux`, as that uses a different provider component.
 */
export default abstract class WrapperWithIntl<
  C extends React.ComponentType<any>,
  P extends React.ComponentProps<C> = React.ComponentProps<C>
> extends Wrapper<C, P> {
  /**
   * The props to be passed to `react-intl`'s `IntlProvider` component
   */
  protected abstract intlProviderProps: Partial<IntlConfig>;

  /**
   * Returns the merged `IntlProvider` props
   */
  protected get mergedIntlProviderProps() {
    return {
      defaultLocale: "en",
      locale: navigator.language,
      messages: {},
      ...this.intlProviderProps,
    };
  }

  /**
   * The component that wraps the component you're testing
   */
  protected WrappingComponent: React.FC = ({ children }) => (
    <IntlProvider {...this.mergedIntlProviderProps}>{children}</IntlProvider>
  );
}
