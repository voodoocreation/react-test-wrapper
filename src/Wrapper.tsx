import { ComponentType, mount, render, shallow } from "enzyme";
import * as React from "react";

export default class Wrapper<
  C extends React.ComponentType<any>,
  P extends React.ComponentProps<C> = React.ComponentProps<C>
> {
  // Properties that are implemented in classes that extend Wrapper
  protected WrappingComponent: ComponentType<any> | undefined = undefined;

  // Properties that remain throughout instance lifecycle
  protected defaultChildren: React.ReactNode;
  protected defaultProps: Partial<P> = {};

  // Scenario-specific properties that are cleared whenever shallow/mount/render are called
  protected scenarioChildren: React.ReactNode;
  protected scenarioProps: Partial<P> = {};

  // Properties to be accessed via getters after mounting
  protected mergedProps: Partial<P> = {};

  constructor(protected readonly Component: C) {}

  public get props() {
    return this.mergedProps;
  }

  public withDefaultChildren = (children: React.ReactNode) => {
    this.defaultChildren = children;

    return this;
  };

  public withDefaultProps = (props: Partial<P>) => {
    this.defaultProps = props;

    return this;
  };

  public withChildren = (children: React.ReactNode) => {
    this.scenarioChildren = children;

    return this;
  };

  public withProps = (props: Partial<P>) => {
    this.scenarioProps = props;

    return this;
  };

  public mount = () => {
    this.beforeMount();

    const props = this.defineProps() as P;
    const wrapper = mount<P>(<this.Component {...props} />, {
      wrappingComponent: this.WrappingComponent,
    });

    this.reset();

    return wrapper;
  };

  public render = () => {
    this.beforeMount();

    const props = this.defineProps() as P;
    const wrapper = render<P, React.ComponentState>(
      this.WrappingComponent ? (
        <this.WrappingComponent>
          <this.Component {...props} />
        </this.WrappingComponent>
      ) : (
        <this.Component {...props} />
      )
    );

    this.reset();

    return wrapper;
  };

  public shallow = () => {
    this.beforeMount();

    const props = this.defineProps() as P;
    const wrapper = shallow<P>(
      this.WrappingComponent ? (
        <this.WrappingComponent>
          <this.Component {...props} />
        </this.WrappingComponent>
      ) : (
        <this.Component {...props} />
      )
    );

    this.reset();

    return wrapper;
  };

  protected beforeMount = () => {
    // Implement this method when extending to define properties to pass to `WrappingComponent`
  };

  protected defineProps = () => {
    this.mergedProps = {
      ...this.defaultProps,
      ...this.scenarioProps,
      children: this.scenarioChildren || this.defaultChildren,
    };

    return this.mergedProps;
  };

  protected reset() {
    this.scenarioChildren = undefined;
    this.scenarioProps = {};
  }
}
