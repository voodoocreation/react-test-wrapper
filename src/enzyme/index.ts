// istanbul ignore file
import Wrapper from "./Wrapper";
import WrapperWithIntl from "./WrapperWithIntl";
import WrapperWithRedux from "./WrapperWithRedux";

export { Wrapper, WrapperWithIntl, WrapperWithRedux };
