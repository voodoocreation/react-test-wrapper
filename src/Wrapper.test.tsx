import { ReactWrapper, ShallowWrapper } from "enzyme";
import * as React from "react";

import { Dummy } from "../test/Dummy";
import Wrapper from "./Wrapper";

describe("Wrapper", () => {
  describe("when using the different render methods", () => {
    const component = new Wrapper(Dummy).withDefaultProps({
      value: "Default value",
    });

    describe("when using 'shallow'", () => {
      const wrapper = component.shallow();

      it("returns ShallowWrapper", () => {
        expect(wrapper).toBeInstanceOf(ShallowWrapper);
      });

      it("has the correct root node", () => {
        expect(wrapper.instance()).toBeInstanceOf(Dummy);
      });

      it("matches snapshot", () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe("when using 'mount'", () => {
      const wrapper = component.mount();

      it("returns ReactWrapper", () => {
        expect(wrapper).toBeInstanceOf(ReactWrapper);
      });

      it("has the correct root node", () => {
        expect(wrapper.instance()).toBeInstanceOf(Dummy);
      });

      it("matches snapshot", () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe("when using 'render'", () => {
      const wrapper = component.render();

      it("returns Cheerio wrapper", () => {
        expect(wrapper.cheerio).toBe("[cheerio object]");
      });

      it("has the correct root node", () => {
        expect(wrapper.hasClass("Dummy")).toBe(true);
      });

      it("matches snapshot", () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("when using the children API", () => {
    const component = new Wrapper(Dummy).withDefaultChildren(
      <div>Default children</div>
    );

    it("renders default children correctly", () => {
      const wrapper = component.render();

      expect(wrapper.find(".Dummy--children").html()).toBe(
        "<div>Default children</div>"
      );
    });

    it("renders test-specific children correctly", () => {
      const wrapper = component
        .withChildren(<span>Test children</span>)
        .render();

      expect(wrapper.find(".Dummy--children").html()).toBe(
        "<span>Test children</span>"
      );
    });

    it("clears test-specific children after previous test and renders default children again", () => {
      const wrapper = component.render();

      expect(wrapper.find(".Dummy--children").html()).toBe(
        "<div>Default children</div>"
      );
    });
  });

  describe("when using the props API", () => {
    const component = new Wrapper(Dummy).withDefaultProps({
      value: "Default value",
    });

    it("renders with default props correctly", () => {
      const wrapper = component.render();

      expect(component.props).toEqual({ value: "Default value" });
      expect(wrapper.find(".Dummy--value").text()).toBe("Default value");
    });

    it("renders with test-specific props correctly", () => {
      const wrapper = component
        .withProps({
          value: "Test value",
        })
        .render();

      expect(component.props).toEqual({ value: "Test value" });
      expect(wrapper.find(".Dummy--value").text()).toBe("Test value");
    });

    it("clears test-specific props after previous test and uses default props again", () => {
      const wrapper = component.render();

      expect(component.props).toEqual({ value: "Default value" });
      expect(wrapper.find(".Dummy--value").text()).toBe("Default value");
    });
  });
});
