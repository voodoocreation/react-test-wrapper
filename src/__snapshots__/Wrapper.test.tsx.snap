// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper when using the different render methods when using 'mount' matches snapshot 1`] = `
<Dummy
  value="Default value"
>
  <div
    className="Dummy"
  >
    <div
      className="Dummy--value"
    >
      Default value
    </div>
    <div
      className="Dummy--children"
    />
    <button
      className="Dummy--button"
      onClick={[Function]}
    >
      Button
    </button>
  </div>
</Dummy>
`;

exports[`Wrapper when using the different render methods when using 'render' matches snapshot 1`] = `
<div
  class="Dummy"
>
  <div
    class="Dummy--value"
  >
    Default value
  </div>
  <div
    class="Dummy--children"
  />
  <button
    class="Dummy--button"
  >
    Button
  </button>
</div>
`;

exports[`Wrapper when using the different render methods when using 'shallow' matches snapshot 1`] = `
<div
  className="Dummy"
>
  <div
    className="Dummy--value"
  >
    Default value
  </div>
  <div
    className="Dummy--children"
  />
  <button
    className="Dummy--button"
    onClick={[Function]}
  >
    Button
  </button>
</div>
`;
